lib/resources/aws/aws_security_group.rb

```diff
@@ -41,9 +41,18 @@ def allow_out_only?(criteria = {})
   private
 
   def allow_only(rules, criteria)
+    rules = allow__focus_on_position(rules, criteria)
```

Comment:
Allow_only: I assumed that allow_only should check if there's a single rule allowing access to/from a single criteria. Because the similar security group rules are merged into a single rule, the rule containing multiple security group id's drops through the check for rules.count as that can be one while there are actually multiple security group rules. That's why I check the user_id_group_pairs.count. As a user I would expect a test for allow only on a security group criteria to allow only traffic to/from exact that single security group, which is what it checks for now.

position: ok so if I understand correctly, the position of the rule is not relevant for the security groups? I must admit that it's not completely clear to me what position indicates or what use this has. I assumed it specifies a position in the list of rules, (which is what it tests for obviously) but the position of the rule is only relevant if you know it upfront. So it could be used to check of a rule has changed position or something. If you would actually expect the rules to be in a particular position as created in the aws interface, that could possibly fail if you use security-groups with similar rules (because they are merged together). I just implemented it in the security group check because it was already there. 

I am perfectly willing to remove/change anything, but maybe you could make it a little more clear what exactly the reuirements for allow_only are and what position implies? If it's just that I'm not getting it that's also fine. Then I'll just remove the "superfluous" code and that's also fine with me. I will have to take some time trying to understand the code a little better then.

Reply:
Ahh, the key fact that "multiple rules that only target security groups get combined" breaks a lot of the assumptions behind the way that allow_only works.  

Indeed, so long as AWS isn't combining or rearranging rules on its own, position is a (potentially) useful way of isolating an individual rule, and then applying exact matching.  But since AWS is re-arranging things, something like what you have added is needed.  Makes me wonder if ipv4/6 range is also being collapsed / expanded.

So, I'm OK with the code, but I think some comments explaining why the `security_group` criterion requires special handling would be great.

Reply:
>  Makes me wonder if ipv4/6 range is also being collapsed / expanded.

I run this through the debugger and it is indeed the case. When multiple similar rules are created they are collapsed into one ip_permission hash with multiple ip_ranges / ipv_6_ranges /   user_id_group_pairs

 >Indeed, so long as AWS isn't combining or rearranging rules on its own, position is a (potentially) useful way of isolating an individual rule, and then applying exact matching.

Could it be possible that the position check was initially based on this comment:  https://github.com/inspec/inspec/issues/2574#issuecomment-369775428 ?

Which I think must be a confusion between security groups, which don't allow 'deny' rules and don't depend on position:
see the second point here:
https://docs.aws.amazon.com/vpc/latest/userguide/VPC_SecurityGroups.html#VPCSecurityGroups

and network ACL's that do allow 'deny' rules and also depend on position:
https://docs.aws.amazon.com/vpc/latest/userguide/vpc-network-acls.html

I think some kind of check for exact rules that searches all rules and sees if there's an exact match might be more appropriate.

Reply:
I have now tested this and the position matching actually fails when two similar rules (only different in target) are before the position.

Reply:
To clarify something about `position`, and the intent of `allow_*` in general: the matcher is not intended to emulate the evaluated behavior of AWS security groups, but only to examine the data structure that is returned by the `describe_security_groups` API call.  In that context, having a `position` element makes sense, as you might be interested in checking a rule located at a particular index.  

Having AWS expand / collapse rules breaks that badly, and that should likely be addressed but it's outside the scope of this PR.

Reply:
I've opened #3602 to address concerns about position; let's not let that block this PR.